File: packages/gbnf-experiment/data/20260909T020947Z_ffb57550/ported_implementation/src/grammar-graph/grammar-graph-types.ts
import { validateNonEmpty } from '../utils/validate-non-empty.js';
import type { GraphPointer } from './graph-pointer.js';
import type { Pointers } from './pointers.js';
import type { RuleRef } from './rule-ref.js';
import { RuleType } from './rule-type.js';

export type Colorize = (text: string | number, color: string) => string;

export interface PrintOpts {
  pointers?: Pointers;
  colorize: Colorize;
  showPosition?: boolean;
}

export type Range = [number, number];

export interface RuleChar {
  type: RuleType.CHAR;
  value: (number | Range)[];
}

export interface RuleCharExclude {
  type: RuleType.CHAR_EXCLUDE;
  value: (number | Range)[];
}

export interface RuleEnd {
  type: RuleType.END;
}

export const ruleChar = (value: (number | Range)[]): RuleChar => ({
  type: RuleType.CHAR,
  value: validateNonEmpty([...value]),
});

export const ruleCharExclude = (
  value: (number | Range)[],
): RuleCharExclude => ({
  type: RuleType.CHAR_EXCLUDE,
  value: validateNonEmpty([...value]),
});

export const ruleEnd = (): RuleEnd => ({ type: RuleType.END });

export type UnresolvedRule = RuleChar | RuleCharExclude | RuleRef | RuleEnd;

// RuleRefs should never be exposed to the end user.
export type ResolvedRule = RuleCharExclude | RuleChar | RuleEnd;

// ValidInput can either be a string, or a number indicating a code point.
// It CANNOT be a number representing a number; a number intended as input (like "8")
// should be passed in as a string.
export type ValidInput = string | number | number[];

export type ResolvedGraphPointer = GraphPointer<ResolvedRule>;

File: proposed-projects/round-trip-experiment/reverse/20260909T181751Z_00b43b48/20260910T095901Z_ef14bd50/ported_implementation/gbnf/grammar_graph/grammar_graph_types.py
from typing import TYPE_CHECKING, Callable, List, Union

from ..utils.validate_non_empty import validate_non_empty
from .rule_ref import RuleRef
from .rule_type import RuleType

if TYPE_CHECKING:
    from .pointers import Pointers

__all__ = ["RuleRef", "RuleType"]

Colorize = Callable[[Union[str, int], str], str]


class PrintOpts:
    def __init__(
        self,
        colorize: Colorize,
        pointers: "Pointers | None" = None,
        show_position: "bool | None" = None,
    ):
        self.pointers = pointers
        self.colorize = colorize
        self.show_position = show_position


# A range is a two element list of code points, inclusive on both ends.
Range = List[int]


class Rule:
    type: RuleType

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, (Rule, RuleRef)):
            return NotImplemented
        return type(self) is type(other) and getattr(self, "value", None) == getattr(
            other, "value", None
        )

    def __ne__(self, other: object) -> bool:
        result = self.__eq__(other)
        if result is NotImplemented:
            return result
        return not result


class RuleWithListOfIntsOrRanges(Rule):
    def __init__(self, value: List[Union[int, Range]]):
        # rules coming in are mutated as ranges and alternates are folded into them;
        # take a copy so that callers holding onto the original list are unaffected.
        self.value = list(validate_non_empty(value))

    def __repr__(self) -> str:
        return f"{type(self).__name__}({self.value!r})"


class RuleChar(RuleWithListOfIntsOrRanges):
    def __init__(self, value: List[Union[int, Range]]):
        self.type = RuleType.CHAR
        super().__init__(value)


class RuleCharExclude(RuleWithListOfIntsOrRanges):
    def __init__(self, value: List[Union[int, Range]]):
        self.type = RuleType.CHAR_EXCLUDE
        super().__init__(value)


class RuleEnd(Rule):
    def __init__(self) -> None:
        self.type = RuleType.END

    def __repr__(self) -> str:
        return "RuleEnd()"


UnresolvedRule = Union[RuleChar, RuleCharExclude, RuleRef, RuleEnd]

# RuleRefs should never be exposed to the end user.
ResolvedRule = Union[RuleCharExclude, RuleChar, RuleEnd]

# ValidInput can either be a string, or a number indicating a code point.
# It CANNOT be a number representing a number; a number intended as input (like "8")
# should be passed in as a string.
ValidInput = Union[str, int, List[int]]
